File: server/utils/request_categories.js
const canonicalCategories = [
  "Groceries & Food",
  "Fashion & Clothing",
  "Electronics & Mobile Accessories",
  "Beauty & Personal Care",
  "Home, Kitchen & Appliances",
  "Health & Pharmacy",
  "Books & Stationery",
  "Baby & Kids",
  "Sports & Fitness",
  "Gifts & Lifestyle",
  "Other",
];

const legacyToCanonical = {
  Groceries: "Groceries & Food",
  "Food & Beverages": "Groceries & Food",
  Clothing: "Fashion & Clothing",
  Electronics: "Electronics & Mobile Accessories",
  Health: "Health & Pharmacy",
  Sports: "Sports & Fitness",
  Books: "Books & Stationery",
};

const aliasesByCanonical = {
  "Groceries & Food": ["Groceries & Food", "Groceries", "Food & Beverages"],
  "Fashion & Clothing": ["Fashion & Clothing", "Clothing"],
  "Electronics & Mobile Accessories": [
    "Electronics & Mobile Accessories",
    "Electronics",
  ],
  "Health & Pharmacy": ["Health & Pharmacy", "Health"],
  "Books & Stationery": ["Books & Stationery", "Books"],
  "Sports & Fitness": ["Sports & Fitness", "Sports"],
};

function normalizeCategory(category) {
  const trimmed = String(category || "").trim();
  if (!trimmed) return "";
  return legacyToCanonical[trimmed] || trimmed;
}

function normalizeCategories(categories) {
  if (!Array.isArray(categories)) return [];
  return [...new Set(categories.map(normalizeCategory).filter(Boolean))];
}

function categoryAliases(category) {
  const canonical = normalizeCategory(category);
  if (!canonical) return [];
  return aliasesByCanonical[canonical] || [canonical];
}

function categoriesWithAliases(categories) {
  return [
    ...new Set(
      normalizeCategories(categories).flatMap((category) =>
        categoryAliases(category),
      ),
    ),
  ];
}

function categoryMatches(profileCategories, requestCategory) {
  const profileValues = new Set(categoriesWithAliases(profileCategories));
  return categoryAliases(requestCategory).some((category) =>
    profileValues.has(category),
  );
}

module.exports = {
  canonicalCategories,
  normalizeCategory,
  normalizeCategories,
  categoryAliases,
  categoriesWithAliases,
  categoryMatches,
};

File: lib/core/config/request_categories.dart
abstract final class RequestCategories {
  static const groceriesFood = 'Groceries & Food';
  static const fashionClothing = 'Fashion & Clothing';
  static const electronicsMobileAccessories =
      'Electronics & Mobile Accessories';
  static const beautyPersonalCare = 'Beauty & Personal Care';
  static const homeKitchenAppliances = 'Home, Kitchen & Appliances';
  static const healthPharmacy = 'Health & Pharmacy';
  static const booksStationery = 'Books & Stationery';
  static const babyKids = 'Baby & Kids';
  static const sportsFitness = 'Sports & Fitness';
  static const giftsLifestyle = 'Gifts & Lifestyle';
  static const other = 'Other';

  static const all = [
    groceriesFood,
    fashionClothing,
    electronicsMobileAccessories,
    beautyPersonalCare,
    homeKitchenAppliances,
    healthPharmacy,
    booksStationery,
    babyKids,
    sportsFitness,
    giftsLifestyle,
    other,
  ];

  static const homeShortcuts = [
    electronicsMobileAccessories,
    fashionClothing,
    groceriesFood,
    beautyPersonalCare,
    homeKitchenAppliances,
  ];

  static const _legacyLabels = {
    'Groceries': groceriesFood,
    'Food & Beverages': groceriesFood,
    'Clothing': fashionClothing,
    'Electronics': electronicsMobileAccessories,
    'Health': healthPharmacy,
    'Sports': sportsFitness,
    'Books': booksStationery,
  };

  static String normalize(String category) {
    final trimmed = category.trim();
    return _legacyLabels[trimmed] ?? trimmed;
  }

  static List<String> normalizeList(Iterable<String> categories) {
    return categories
        .map(normalize)
        .where((category) => category.isNotEmpty)
        .toSet()
        .toList();
  }
}
